### PR TITLE
initial commit of the branch SSL_comparison

### PR DESCRIPTION
**Pull request by JulietteMoreau**
For my SSL benchmark review of next Thursday!

### CODE REVIEW

File: collections/_posts/2024-12-02-SSL_comparison.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+* Implementation of a real life hyperparameters optimization.
+* Comparison of semi-supervised and self-supervised learning.
+
+# Note
+
+Code is available here: https://github.com/tufts-ml/SSL-vs-SSL-benchmark
+
+# Introduction
+
+In medical domain there is often a lack of annotated data for deep learning model training while there is a lot of unlabeled data in data health records. Semi-supervised and self-supervised learning are methods developed to take advantage of these unlabeled images to improve classification accuracy. The first one train classifiers jointly with two loss terms, while the second is a two-stages approach, the first to learn deep representation and the second to fine-tune the classifier. However, the two methods are rarely compared. The first question they try to answer is: **Which recent semi- or self-supervised methods are likely to be most effective?** \
```
Comment (Clarax99) [inline collections/_posts/2024-12-02-SSL_comparison.md:26]: - In **the** medical domain
- "while there is a lot of unlabeled data in data health records" → Awkward phrasing. Suggested: "while a large amount of unlabeled data exists in health records."
- The first one **trains** classifiers jointly
- the second is a **two-stage** approach
- General comment : maybe you could gain in lisibility with bullet points for Semi and SSL such as : 

- Semi : blablabla introductive content
- SSL: introductive content with two-stage approach

1. Pretraining
2. Finetuning


Suggestion of rephrasing of chatGPT (2ème review pour le prix d'une hihi): 
In the medical domain, annotated data for deep learning is often scarce, while a large amount of unlabeled data exists in health records. Semi-supervised and self-supervised learning methods have been developed to leverage this unlabeled data and improve classification accuracy. The former trains classifiers jointly with two loss terms, while the latter follows a two-stage approach: first, learning deep representations, and second, fine-tuning the classifier. However, these two methods are rarely compared.

File: collections/_posts/2024-12-02-SSL_comparison.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+But those methods are very sensitive to hyperparameters, and benchmarks often don't consider this (either no hyperparameters tuning or tuning with a huge labeled validation set, bigger than the training set, which is not realistic). In this paper, they try to take that in account by answering the question: **Given limited available labeled data and limited compute, is hyperparameter tuning worthwhile?**
```
Comment (Clarax99) [inline collections/_posts/2024-12-02-SSL_comparison.md:27]: benchmarks often don't consider **this** : Not very clear imo, it could be more explicit. Maybe you could add another sentence instead of the note in parentheses

File: collections/_posts/2024-12-02-SSL_comparison.md [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+# Methods
+
+## Definition of SSL
+
+A unified perspective for different learning paradigms is presented with two available datasets: $$L$$ which is a small labeled dataset and $$U$$ the unlabeled dataset with $$ \lvert U \rvert >> \lvert L \rvert $$. The learning process can be written as: 
+
+$$ v^*, w^* ← argmin_{v,w} \sum_{x,y \in L} \lambda^L l^L(y, g_w(f_v(x))) + \sum_{x \in U} \lambda^U l^U(x, f_v,g_w) $$ 
+
+where $$l^L$$ and $$l^U$$ are the loss functions linked to the labeled and unlabeled datasets and $$\lambda^L$$ and $$\lambda_{U}$$ the associated weights. In addition, $$f_v(\cdot)$$  denotes a neural network backbone with parameters $$v$$ that produces an embedding of the input images and $$g_w(\cdot)$$ denotes a final linear softmax classification layer with parameters $$w$$.
```
Comment (Clarax99) [inline collections/_posts/2024-12-02-SSL_comparison.md:37]: $$\lambda_{U}$$ -->  $$\lambda^{U}$$ for uniformity

File: collections/_posts/2024-12-02-SSL_comparison.md [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+| Method          | $$\lambda^L$$ | $$\lambda^U$$ |
+|-----------------|---------------|---------------|
+| Supervised      | 1             | 0             |
+| Semi-supervised | >0            | >0            |
+| Self-supervised | 0 then 1      | 1 then 0      |
```
Comment (Clarax99) [inline collections/_posts/2024-12-02-SSL_comparison.md:43]: "Self-supervised | 0 then 1 | 1 then 0" → Could be clearer. Suggested: "Self-supervised | Pretraining (0) → Fine-tuning (1) | Pretraining (1) → Fine-tuning (0)." 
I agree with Chatgpt on this point but i can understand it doesn't fit nicely, so up to you

File: collections/_posts/2024-12-02-SSL_comparison.md [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+## Compared methods
+
+**Supervised learning**
+
+Only using the labeled $$(L)$$ dataset.
+
+* Sup: to denote a classifier trained with classical multiclass cross entropy loss.
+* MixUp: also multiclass cross entropy, but with mixup data augmentation (creating new training samples by mixing two samples from the original dataset).
+* SupCon: using a supervised contrastive learning loss (pull together samples belonging to the same class in the embedded space while pushing apart from samples of other classes).
```
Comment (Clarax99) [inline collections/_posts/2024-12-02-SSL_comparison.md:53]: I am a bit confused by this subsection, I imagine this is both the methods and their notation for the compared supervised methods, maybe it is lacking an introductive sentence (which you did in the next subsections)

File: collections/_posts/2024-12-02-SSL_comparison.md [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+| Self-supervised | 0 then 1      | 1 then 0      |
+
+## Compared methods
+
+**Supervised learning**
+
+Only using the labeled $$(L)$$ dataset.
+
+* Sup: to denote a classifier trained with classical multiclass cross entropy loss.
+* MixUp: also multiclass cross entropy, but with mixup data augmentation (creating new training samples by mixing two samples from the original dataset).
```
Comment (Clarax99) [inline collections/_posts/2024-12-02-SSL_comparison.md:52]: "creating new training samples by mixing two samples from the original dataset" → Suggested: "by linearly combining two training samples and their corresponding labels."

File: collections/_posts/2024-12-02-SSL_comparison.md [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+* SupCon: using a supervised contrastive learning loss (pull together samples belonging to the same class in the embedded space while pushing apart from samples of other classes).
+
+**Semi-supervised learning**
+
+They selected 6 methods of the recent years to cover a wide spectrum of strategies and cost, to take advantage of the unlabeled $$(U)$$ and labeled $$(L)$$ datasets jointly.
+
+* Pseudo-label: sets the label to unlabeled data to the class which has the maximum predicted probability.
```
Comment (Clarax99) [inline collections/_posts/2024-12-02-SSL_comparison.md:59]: "sets the label to unlabeled data to the class..." → Suggested: "assigns labels to unlabeled data based on the class with the highest predicted probability."

File: collections/_posts/2024-12-02-SSL_comparison.md [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+* Mean teacher: teacher weights are updated as exponential moving average of student weights after each sample, either the sample used is labeled or not.
+* MixMatch: uses mixup data augmentation to combine label and unlabeled data to compute labels for this new input images.
```
Comment (Clarax99) [inline collections/_posts/2024-12-02-SSL_comparison.md:61]: combine **labeled** and unlabeled data

File: collections/_posts/2024-12-02-SSL_comparison.md [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+* FixMatch: combination of pseudo-label method to set a label to unlabeled images if the prediction confidence is high enough with strong data augmentation, which prediction has to match the label set previously.
+* FlexMatch: adding a curriculum notion to FixMatch to consider different learning status and learning difficulties of different classes, basically adjusting the threshold of confidence to define a class to an unlabeled image.
+* CoMatch: learns jointly class probability and low-dimensional embeddings to impose a smoothness constraint and improve the pseudo-labels, while those last  regularize the structure of the embeddings through graph based contrastive learning.
+
+**Self-supervised learning**
+
+In addition, 7 self-supervised methods are compared:
+
+* SimCLR: contrastive learning self-supervised method based on data augmentation with a learning nonlinear transformation between the representation and contrastive loss to improve the learned representations.
```
Comment (Clarax99) [inline collections/_posts/2024-12-02-SSL_comparison.md:70]: "with a learning nonlinear transformation..." → Awkward phrasing. Suggested: "using a learnable nonlinear transformation..."

File: collections/_posts/2024-12-02-SSL_comparison.md [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+* MOCO v2: for momentum contrastive, which uses a dynamic dictionary with a queue and a moving-averaged encoder that facilitates contrastive unsupervised learning.
+* SwAV: contrastive learning that does not compare features directly but comparing different augmentation of the same images.
```
Comment (Clarax99) [inline collections/_posts/2024-12-02-SSL_comparison.md:72]: "...but comparing different augmentation of the same images." → Correct: "...but compares different augmentations of the same image."

File: collections/_posts/2024-12-02-SSL_comparison.md [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+Here are some images examples.
+
+![](/collections/images/SSL_comparison/datasets_example.jpg)
+
+# Experimental setup
+
+For the low resolution images ResNet-18 is used, WideResNet-28-2 is used for TMED-2 and ResNet-18 or 50 are tested for AIROGS dataset. \
+A maximum of 200 epochs is performed and training is stopped if the balanced accuracy plateaus for 20 consecutive epochs. 
+
+Hyperparameters (learning rate, weight decay and unlabeled loss weight, in addition to the parameters specific for each method) optimization is made for each method and dataset. To fit with real life constraints, they fixed a certain number of hours with a NVIDIA A100 GPU for each hyperparameter optimization (25h for PathMNIST, 50h for TissueMNIST and 100h for TMED-2 and AIROGS). Within this fixed budget, a random search of hyperparameters is done, tacking the best with validation set.
```
Comment (Clarax99) [inline collections/_posts/2024-12-02-SSL_comparison.md:102]: **taking** the best **using the** validation set.

File: collections/_posts/2024-12-02-SSL_comparison.md [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+
+# Results
+
+Five separate trials were performed and mean balanced accuracy is calculated each 30 or 60 minutes on validation and test sets.
+
+Balanced accuracy on test set over time is represented for each dataset and method. As there is a rough monotonic improvement in test performance over time despite using a realistic-size validation set, they conclude that checkpoint selection and hyperparameters optimization can be effective with a realistically-sized validation set.
```
Comment (Clarax99) [inline collections/_posts/2024-12-02-SSL_comparison.md:108]: "As there is a rough monotonic improvement in test performance..." → Suggested: " As a roughly monotonic improvement in test performance is observed..."

File: collections/_posts/2024-12-02-SSL_comparison.md [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+
+![](/collections/images/SSL_comparison/accuracy_evolution.jpg)
+
+All the methods are compared and none of them clearly stands out from the other depending on the dataset. So they measure the balanced accuracy relative gain over the best supervised method for each dataset. MixMatch represent the best overall choice.
```
Comment (Clarax99) [inline collections/_posts/2024-12-02-SSL_comparison.md:112]: MixMatch **represents** the best overall choice.

File: collections/_posts/2024-12-02-SSL_comparison.md [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+![](/collections/images/SSL_comparison/results_comparison.jpg)
+
+The tuning strategy, even though it reserves data for a validation set, is competitive in comparison with transferring the hyperparameters from another dataset (natural images or medical images) as it may have been done in other benchmark papers.
+
+![](/collections/images/SSL_comparison/hyperparameters_strategy_evaluation.jpg)
+
+
+# Discussion
+
+There is a real gain in performance of performing SSL, but it depends on the dataset (less gain on TMED-2). However, they insist on the importance of the selection of the hyperparameters and proper evaluation protocol depending on specific needs. Especially MixMatch requires cautious hyperparameters tuning when applied to a new dataset.
```
Comment (Clarax99) [inline collections/_posts/2024-12-02-SSL_comparison.md:123]: - "There is a real gain in performance of performing SSL" → Suggested: "SSL provides a tangible performance gain..."
- "less gain on TMED-2" → Unclear. Suggested: "a smaller performance improvement is observed for TMED-2."
- "MixMatch requires cautious hyperparameters tuning" → Suggested: "MixMatch requires careful hyperparameter tuning..."

**Review by Clarax99 — APPROVED** [review]
The content is structured and clear, sometimes the chosen phrasing is not the best but understandable.
I let you some remarks for some minor grammar corrections and rephrasing suggestions, up to you to take them into account. Sorry for the late review!

File: collections/_posts/2024-12-02-SSL_comparison.md [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+* Implementation of a real life hyperparameters optimization.
+* Comparison of semi-supervised and self-supervised learning.
+
+# Note
+
+Code is available here: https://github.com/tufts-ml/SSL-vs-SSL-benchmark
+
+# Introduction
+
+In the medical domain there is often a lack of annotated data for deep learning model training while there is a lot of unlabeled data in data health records. Semi-supervised and self-supervised learning are methods developed to take advantage of these unlabeled images to improve classification accuracy. The first one trains classifiers jointly with two loss terms, while the second is a two-stage approach, the first to learn deep representation and the second to fine-tune the classifier. However, the two methods are rarely compared. The first question they try to answer is: **Which recent semi- or self-supervised methods are likely to be most effective?** \
```
Comment (nathanpainchaud) [inline collections/_posts/2024-12-02-SSL_comparison.md:26]: Repeated word:
```suggestion
In the medical domain there is often a lack of annotated data for deep learning model training while there is a lot of unlabeled data in health records. Semi-supervised and self-supervised learning are methods developed to take advantage of these unlabeled images to improve classification accuracy. The first one trains classifiers jointly with two loss terms, while the second is a two-stage approach, the first to learn deep representation and the second to fine-tune the classifier. However, the two methods are rarely compared. The first question they try to answer is: **Which recent semi- or self-supervised methods are likely to be most effective?** \
```

File: collections/_posts/2024-12-02-SSL_comparison.md [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+
+# Highlights
+
+* Benchmark of SSL methods on medical images.
+* Implementation of a real life hyperparameters optimization.
+* Comparison of semi-supervised and self-supervised learning.
+
+# Note
+
+Code is available here: https://github.com/tufts-ml/SSL-vs-SSL-benchmark
```
Comment (nathanpainchaud) [inline collections/_posts/2024-12-02-SSL_comparison.md:22]: It would convenient for the link to be clickable :wink: 
```suggestion
Code is available here: [https://github.com/tufts-ml/SSL-vs-SSL-benchmark](https://github.com/tufts-ml/SSL-vs-SSL-benchmark)
```

File: collections/_posts/2024-12-02-SSL_comparison.md [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+
+$$ v^*, w^* ← argmin_{v,w} \sum_{x,y \in L} \lambda^L l^L(y, g_w(f_v(x))) + \sum_{x \in U} \lambda^U l^U(x, f_v,g_w) $$ 
+
+where $$l^L$$ and $$l^U$$ are the loss functions linked to the labeled and unlabeled datasets and $$\lambda^L$$ and $$\lambda^{U}$$ the associated weights. In addition, $$f_v(\cdot)$$  denotes a neural network backbone with parameters $$v$$ that produces an embedding of the input images and $$g_w(\cdot)$$ denotes a final linear softmax classification layer with parameters $$w$$.
+
+| Method          | $$\lambda^L$$                     | $$\lambda^U$$                       |
+|-----------------|-----------------------------------|-------------------------------------|
+| Supervised      | 1                                 | 0                                   |
+| Semi-supervised | >0                                | >0                                  |
+| Self-supervised | pretraining 0() --> finetuning (1)  pretraining (1) --> finetuning (0)  |
```
Comment (nathanpainchaud) [inline collections/_posts/2024-12-02-SSL_comparison.md:43]: Fix the layout:
```suggestion
| Self-supervised | pretraining (0) --> finetuning (1) | pretraining (1) --> finetuning (0) |
```

File: collections/_posts/2024-12-02-SSL_comparison.md [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+
+## Compared methods
+
+**Supervised learning**
+
+Three supervised methods are used as reference to compare with SSL, those trainings only use the labeled $$(L)$$ dataset.
```
Comment (nathanpainchaud) [inline collections/_posts/2024-12-02-SSL_comparison.md:49]: For clarity:
```suggestion
Three supervised methods are used as reference to compare with SSL. These supervised methods only use the labeled $$(L)$$ dataset.
```

**Review by nathanpainchaud — CHANGES_REQUESTED** [review]
A few things that caught my eye during the review. Otherwise very clear review, good job :slightly_smiling_face:
